docs/sync-storage-handbook/src/data-model.md

```diff
@@ -0,0 +1,54 @@
+# Sync data model
+
+- Introduce high-level concepts here, and elaborate on the details elsewhere?
+
+## Data classes
+
+All synced data types fit into one of three broad categories.
+
+- Are there others we should think about?
+- Would a table comparing the different characteristics be useful?
+- These classifications aren't rigid; it'd be nice to find a way to describe that better.
+
+### Hierarchical data
+
+Bookmarks are an example of hierarchical data, or **trees**. Records in tree collections are interdependent: parents have pointers to their children, and children back to their parents.
+
+Hierarchical data present a problem for syncing, because records can be downloaded in any order. Additionally, some changes require multiple records to be uploaded in lockstep. Missing or incomplete records lead to problems like missing parents and children, and parent-child disagreements, which clients must resolve to ensure the server data remain consistent.
+
+### Append-only data
+
+History is an example of append-only, or **log** data. Since all entries are distinct, there are no conflicts: two visits to the same page on two different devices are still two distinct visits. These records are independent, and can be synced in any order.
+
+Log data are the easiest to sync, but the problem is _volume_. We currently limit history to the last 20 visits per page, cap initial syncs to the last 30 days or 5,000 pages, and expire records that aren't updated on the server after 60 days. These limitations are for efficiency as much as for historical reasons. Clients don't need to process thousands of entries on each sync, and the server avoids bloated indexes for large collections. Unfortunately, this is also a form of data loss, as the server never has the complete history.
+
+### Semistructured data
+
+Logins, addresses, and credit cards are examples of semistructured, or **document** data. Like log data, document data are independent, and can be synced in any order relative to each other. However, they _can_ conflict if two clients change the same field.
+
+Engines that implement three-way merging support per-field conflict resolution, since they store the value that was last uploaded to the server. Engines that don't resolve conflicts at the record level, based on the timestamp.
```

Comment:
nit: on my first read I mis-parsed this as talking about "engines that don't resolve conflicts at the record level" rather than "engines that don't implement three-way merging". It might be clearer as something like "Engines that don't will instead resolve conflicts...".

Reply:
Do we have any such engines in our current suite, or do we intend to in the future?